### PR TITLE
sink(ticdc): change the directory of storage sink only when ddl event occurs

### LINKED ISSUE #8890: Change the directory of storage sink only when ddl event occurs

### LINKED ISSUE #8891: Make storage sink support database level DDL

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #8890, close #8891

### What is changed and how it works?
1. Change the directory of storage sink only when ddl event occurs.
2. Make storage sink support database level DDL.

- File path changes:
1. table schema file: 
from:
`{scheme}://{prefix}/{schema}/{table}/{table-version-separator}/schema.json`
to:
`{scheme}://{prefix}/{schema}/{table}/meta/schema_{tso}_{schema-crc32-hash}.json`
2. database schema:
`{scheme}://{prefix}/{schema}/meta/schema_{tso}_{schema-crc32-hash}.json`
3. index file:
from:
`{scheme}://{prefix}/{schema}/{table}/{table-version-separator}/{partition-separator}/{date-separator}/CDC.index
`
to:
`{scheme}://{prefix}/{schema}/{table}/{table-version-separator}/{partition-separator}/{date-separator}/meta/CDC.index`

#### Tests 

 - Unit test
 - Integration test

##### Will it cause performance regression or break compatibility?
- The schema-related storage path structure in storage sink has changed

##### Do you need to update user documentation, design documentation or monitoring documentation?
- need to update user documentation and design documentation

### Release note 

```release-note
`Fixed the table version directory written by storage sink may change while cdc restart or table scheduling`.
`Support database level DDL in storage sink.`
```

### CODE REVIEW

**Review by nongfushanquan — COMMENTED** [review]
pls add comments to describe  the layout of the data and schema

**Review by nongfushanquan — APPROVED** [review]

**Comment by CharlesCheung96** [comment]
test infra verified: 1669199

File: cdc/sink/ddlsink/cloudstorage/cloud_storage_ddl_sink_test.go [hunk 1]
```diff
@@ -63,12 +63,12 @@ func TestWriteDDLEvent(t *testing.T) {
 			},
 		},
 	}
-	tableDir := path.Join(parentDir, "test/table1/100")
-	os.MkdirAll(tableDir, 0o755)
+	tableDir := path.Join(parentDir, "test/table1/meta/")
+	// os.MkdirAll(tableDir, 0o755)
```
Comment (hi-rustin) [inline cdc/sink/ddlsink/cloudstorage/cloud_storage_ddl_sink_test.go:67]: Please remove commented code.

File: cmd/storage-consumer/main.go [hunk 2]
```diff
@@ -563,24 +544,112 @@ func (c *consumer) syncExecDMLEvents(
...
+
+func (c *consumer) clenaupTableDef(key SchemaPathKey) {
+	// if tableDefs, ok := c.tableDefMap[key.getKey()]; ok {
+	// 	for k := range tableDefs {
+	// 		if k < key.TableVersion {
+	// 			delete(tableDefs, k)
+	// 		}
+	// 	}
+	// }
+	// Do nothin for now. we may need to cleanup tableDefMap in the future.
```
Comment (hi-rustin) [inline cmd/storage-consumer/main.go:652]: Please remove commented code.

File: cmd/storage-consumer/main_test.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/pkg/leakutil"
+	"go.uber.org/zap"
+)
+
+func TestMain(m *testing.M) {
+	leakutil.SetUpLeakTest(m)
+}
+
+func TestSchemaPathKey(t *testing.T) {
```
Comment (hi-rustin) [inline cmd/storage-consumer/main_test.go:29]: Usually, we don't put any tests in `mac_test.go`.

File: pkg/sink/cloudstorage/path_test.go [hunk 4]
```diff
@@ -228,3 +233,46 @@ func TestGenerateDataFilePathWithIndexFile(t *testing.T) {
...
+		{"invalid path", "meta/schema1/schema_123_0123456789.json", false},
+	}
+
+	for _, tt := range tests {
+		require.Equal(t, tt.expect, IsSchemaFile(tt.path),
+			"testCase: %s, path: %v", tt.name, tt.path)
+	}
+}
+
+// func TestGenerateSchemaFilePath(t *testing.T) {
```
Comment (hi-rustin) [inline pkg/sink/cloudstorage/path_test.go:271]: Please remove commented code.

File: pkg/sink/cloudstorage/table_definition_test.go [hunk 5]
```diff
@@ -445,3 +445,7 @@ func TestTableDefinition(t *testing.T) {
 	require.Equal(t, timodel.ActionAddColumn, event.Type)
 	require.Equal(t, uint64(100), event.CommitTs)
 }
+
+func TestTableDefinitionSum32(t *testing.T) {
+	// TODO: add sum32 tests
```
Comment (hi-rustin) [inline pkg/sink/cloudstorage/table_definition_test.go:450]: Maybe we can add it. I guess it is easy to add. At least, we should use `skip()` to ship this unimplemented case.

File: tests/integration_tests/canal_json_storage_partition_table/data/prepare.sql [hunk 6]
```diff
@@ -21,10 +21,11 @@ alter table t1 drop partition p1;
 insert into t1 values (7),(8),(9);
 update t1 set a=a+10 where a=9;
 
-create table t2 (a int primary key);
-ALTER TABLE t1 EXCHANGE PARTITION p3 WITH TABLE t2;
-insert into t2 values (100),(101),(102),(103),(104),(105); /*these values will be replicated to in downstream t2*/
-insert into t1 values (25),(29); /*these values will be replicated to in downstream t1.p3*/
+/* TODO(CharlesCheung): EXCHANGE PARTITION will be supported in the future */
```
Comment (hi-rustin) [inline tests/integration_tests/canal_json_storage_partition_table/data/prepare.sql:24]: Please remove this commented code and open an issuer for it.

**Review by hi-rustin — COMMENTED** [review]
Thanks!

File: pkg/sink/cloudstorage/table_definition_test.go (hunk 5 above)
Comment (CharlesCheung96) [reply pkg/sink/cloudstorage/table_definition_test.go]: https://github.com/pingcap/tiflow/issues/8914

File: tests/integration_tests/canal_json_storage_partition_table/data/prepare.sql (hunk 6 above)
Comment (CharlesCheung96) [reply tests/integration_tests/canal_json_storage_partition_table/data/prepare.sql:24]: https://github.com/pingcap/tiflow/issues/8914

**Review by hi-rustin — APPROVED** [review]

**Pull request merged**